### PR TITLE
Adam optimizer

### PR DESCRIPTION
**Pull request by udibr**
This is part of https://github.com/dmlc/mxnet/issues/619
I've run few tests and it looks ok. But I'm new to mxnet code so please review carefully.

### CODE REVIEW

**Comment by udibr** [comment]
sorry, I had really hard time with git and python (should have been easy.)
I tried to squash all the log messages but failed :-(
at least I hope the correct code is now used

**Comment by piiswrong** [comment]
Looks good.
Have you tried solving the MNIST/mlp.py example with your optimizer?

**Comment by mli** [comment]
many thanks for the PR. please give me a few hours, i would like to do a little bit refactor of the optimizer, mainly for fixing the terminology. the problem is that things are not straightforward when go to the distributed training.

**Comment by udibr** [comment]
It gave me the same results on mnist as SGD.
Adam usually helps in speed and less in quality of result.
To see its speedup you need a deep/recurrent network. I'm trying it now on char-lstm

**Comment by mli** [comment]
i created an issue to discuss it #625

File: python/mxnet/optimizer.py [hunk 1]
```diff
@@ -171,6 +172,130 @@ def update(self, index, weight, grad, state):
...
+        grad : NDArray
+            grad ndarray
+
+        state : NDArray or other objects returned by init_state
+            The auxiliary state used in optimization.
+        """
+        assert(isinstance(weight, NDArray))
+        assert(isinstance(grad, NDArray))
+        if self.lr_scheduler != None:
+            lr = self.lr_scheduler(self.epoch)
```
Comment (mli) [inline python/mxnet/optimizer.py]: could you please change the above to 
```
         if self.lr_scheduler != None:
            lr = self.lr_scheduler(self.num_update)
            self._update_count(index)
```
if this PR is merged after #627

**Comment by tqchen** [comment]
Use ```is not None``` instead of ```!= None```, which is more convention in python

**Comment by tqchen** [comment]
@mli  Can you track this issue, and give further instructions?

File: python/mxnet/optimizer.py [hunk 2]
```diff
@@ -171,6 +172,130 @@ def update(self, index, weight, grad, state):
...
+        grad : NDArray
+            grad ndarray
+
+        state : NDArray or other objects returned by init_state
+            The auxiliary state used in optimization.
+        """
+        assert(isinstance(weight, NDArray))
+        assert(isinstance(grad, NDArray))
+        if self.lr_scheduler is not None:
+            lr = self.lr_scheduler(self.epoch)
```
Comment (mli) [inline python/mxnet/optimizer.py]: (looks like i lost my last comment)

could you please first merge the current master, and then replace the above line by

```python
            lr = self.lr_scheduler(self.num_update)
            self._update_count(index)
```
refer to https://github.com/dmlc/mxnet/blob/master/python/mxnet/optimizer.py#L158

**Comment by tqchen** [comment]
@udibr can you update this PR with the latest master and fixes the comment @mli mentioned?

Thanks!

File: python/mxnet/optimizer.py (hunk 2 above)
Comment (mli) [inline python/mxnet/optimizer.py]: please use `self.num_update` rather than `self.epoch`, which has been already removed 

others are LGTM

**Comment by antinucleon** [comment]
Thanks! Merged

**Pull request merged**